cardano-crypto-praos/src/Cardano/Crypto/VRF/Praos.hs

```diff
@@ -0,0 +1,464 @@
+
+instance FromCBOR SK where
+  fromCBOR = skFromBytes <$> fromCBOR
+
+
+instance Show PK where
+  show = show . pkBytes
+
+instance Eq PK where
+  a == b = pkBytes a == pkBytes b
+
+instance ToCBOR PK where
+  toCBOR = toCBOR . pkBytes
+  encodedSizeExpr _ _ =
+    encodedSizeExpr (\_ -> fromIntegral verKeySizeVRF) (Proxy :: Proxy ByteString)
+
+instance FromCBOR PK where
+  fromCBOR = pkFromBytes <$> fromCBOR
+
+-- | Allocate a Public Key and attach a finalizer. The allocated memory will
+-- not be initialized.
+mkPK :: IO PK
+mkPK = fmap PK $ newForeignPtr finalizerFree =<< mallocBytes verKeySizeVRF
+
+-- | Allocate a Secret Key and attach a finalizer. The allocated memory will
+-- not be initialized.
+mkSK :: IO SK
+mkSK = fmap SK $ newForeignPtr finalizerFree =<< mallocBytes signKeySizeVRF
+
+-- | Allocate a Proof and attach a finalizer. The allocated memory will
+-- not be initialized.
+mkProof :: IO Proof
+mkProof = fmap Proof $ newForeignPtr finalizerFree =<< mallocBytes (certSizeVRF)
+
+proofFromBytes :: ByteString -> Proof
+proofFromBytes bs
+  | BS.length bs /= certSizeVRF
+  = error "Invalid proof length"
+  | otherwise
+  = unsafePerformIO $ do
+      proof <- mkProof
+      withForeignPtr (unProof proof) $ \ptr ->
+        BS.useAsCString bs $ \cstr -> do
+          copyBytes (castPtr ptr) cstr (certSizeVRF)
+      return proof
+
+skFromBytes :: ByteString -> SK
+skFromBytes bs
+  | BS.length bs /= signKeySizeVRF
+  = error "Invalid sk length"
+  | otherwise
+  = unsafePerformIO $ do
+      sk <- mkSK
+      withForeignPtr (unSK sk) $ \ptr ->
+        BS.useAsCString bs $ \cstr -> do
+          copyBytes (castPtr ptr) cstr signKeySizeVRF
+      return sk
+
+pkFromBytes :: ByteString -> PK
+pkFromBytes bs
+  | BS.length bs /= verKeySizeVRF
+  = error "Invalid pk length"
+  | otherwise
+  = unsafePerformIO $ do
+      pk <- mkPK
+      withForeignPtr (unPK pk) $ \ptr ->
+        BS.useAsCString bs $ \cstr -> do
+          copyBytes (castPtr ptr) cstr verKeySizeVRF
+      return pk
+
+-- | Allocate an Output and attach a finalizer. The allocated memory will
+-- not be initialized.
+mkOutput :: IO Output
+mkOutput = fmap Output $ newForeignPtr finalizerFree =<< mallocBytes (fromIntegral crypto_vrf_outputbytes)
+
+-- | Derive a Public/Secret key pair from a seed.
+keypairFromSeed :: Seed -> (PK, SK)
+keypairFromSeed seed =
+  unsafePerformIO $ withForeignPtr (unSeed seed) $ \sptr -> do
+    pk <- mkPK
+    sk <- mkSK
+    withForeignPtr (unPK pk) $ \pkPtr -> do
+      withForeignPtr (unSK sk) $ \skPtr -> do
+        void $ crypto_vrf_keypair_from_seed pkPtr skPtr sptr
+    return (pk, sk)
+
+-- | Derive a Public Key from a Secret Key.
+skToPK :: SK -> PK
+skToPK sk =
+  unsafePerformIO $ withForeignPtr (unSK sk) $ \skPtr -> do
+    pk <- mkPK
+    withForeignPtr (unPK pk) $ \pkPtr -> do
+      void $ crypto_vrf_sk_to_pk pkPtr skPtr
+    return pk
+
+-- | Get the seed used to generate a given Secret Key
+skToSeed :: SK -> Seed
```

Comment:
Do we ever need this?

Reply:
I don't know, I think I added it "for consistency" (i.e., "because we can"). Happy to remove it if we're positively convinced we won't need it.